src/main/dashboardServer/index.ts

```diff
@@ -0,0 +1,109 @@
+import { Server } from 'http';
+import path from 'path';
+
+import { app } from 'electron';
+import express from 'express';
+import { createProxyMiddleware, Options } from 'http-proxy-middleware';
+
+import { proxyWsOpts, proxyOpts, proxyMetaOpts } from './proxyUtils';
+
+import Logging from '@/utils/logging';
+
+type ProxyKeys = '/k8s' | '/pp' | '/api' | '/apis' | '/v1' | '/v3' | '/v3-public' | '/api-ui' | '/meta' | '/v1-*';
+
+type ProxyMap = Record<ProxyKeys, Options>;
+
+const console = Logging.dashboardServer;
+
+/**
+ * Singleton that manages the lifecycle of the Dashboard server.
+ */
+export class DashboardServer {
+  private static instance: DashboardServer;
+
+  private dashboardServer = express();
+  private dashboardApp: Server = new Server();
+  private host = '127.0.0.1';
+  private port = 6120;
+  private api = 'https://127.0.0.1:9443';
+  private proxy: ProxyMap = {
+    '/k8s':          proxyWsOpts(this.api), // Straight to a remote cluster (/k8s/clusters/<id>/)
+    '/pp':           proxyWsOpts(this.api), // For (epinio) standalone API
+    '/api':          proxyWsOpts(this.api), // Management k8s API
+    '/apis':         proxyWsOpts(this.api), // Management k8s API
+    '/v1':           proxyWsOpts(this.api), // Management Steve API
+    '/v3':           proxyWsOpts(this.api), // Rancher API
+    '/v3-public':    proxyOpts(this.api), // Rancher Unauthed API
+    '/api-ui':       proxyOpts(this.api), // Browser API UI
+    '/meta':         proxyMetaOpts(this.api), // Browser API UI
+    '/v1-*':         proxyOpts(this.api), // SAML, KDM, etc
+  };
+
+  private proxies: any = {};
+
+  /**
+   * Checks for an existing instance of Dashboard server.
+   * Instantiate a new one if it does not exist.
+   */
+  public static getInstance(): DashboardServer {
+    if (!DashboardServer.instance) {
+      DashboardServer.instance = new DashboardServer();
+    }
+
+    return DashboardServer.instance;
+  }
+
+  /**
+   * Starts the Dashboard server if one is not already running.
+   */
+  public init() {
+    if (this.dashboardApp.address()) {
+      console.log(`Dashboard Server is already listening on ${ this.host }:${ this.port }`);
+
+      return;
+    }
+
+    Object.entries(this.proxy).forEach(([key, value]) => {
```

Comment:
I attempted this in a way so that we could guarantee that the `proxies` object is typed correctly, but I ran into quite a few issues with `Object.entries` providing few guarantees about keys. There was also an issue with invoking `this.proxies['/v1'].upgrade(req, socket, head);` further down.. 

Not happy about the implementation but typing `proxies` as `any` allowed me to keep moving on this after I exceeded my timebox..

Recommendations are welcome 🙂

Reply:
I was running into similar struggles with typing object methods for this task and I'm left wondering if either the data structure for `proxy` & `proxies` might be incorrect or if the types can be improved..

```
    return Object.fromEntries(Object.entries(proxy).map(([key, options]) => {
      return [key, createProxyMiddleware(options)];
    })) as unknown as Record<ProxyKeys, RequestHandler>;
```

Either way, I'll try to implement what you have recommended here as it seems to get us closer to what we want.